harvester/harvester
# Enhance non-live migratable VMs detection in both webhook and upgrade script

**Problem:**


The part handling non-live-migratable VMs in the upgrade script is a bit hard to read and maintain.

**Solution:**


This PR contains three part:

1. A command line interface called `vm-live-migrate-detector`, which is able to show the "non-live migratable VMs" given a node name as its positional argument. Plus a `--shutdown` flag that could help users shut off those non-migratable VMs directly with the tool at hand. The binary is targeted to replace the respective part in the upgrade shell script.
2. A webhook that does the validation work against Upgrade objects creation. When there are "non-live migratable VMs" running on the cluster, it will reject the request and return the VM name list.
3. A shared library code supporting the above two

**Related Issue:**

#3019 

**Test plan:**


*VMs with node affinities*

1. Prepare a 2-node Harvester cluster in v1.2.1
5. Create two VMs: one is live-migratable, the other is non-live-migratable. To achieve this, we could tweak the node scheduling option for the VMs.
![SCR-20240116-okkg](https://github.com/harvester/harvester/assets/1827717/b8049a2d-4b67-4d2a-afd5-72083014be00)
![SCR-20240116-okdu](https://github.com/harvester/harvester/assets/1827717/bfb6b721-30b3-4c9b-949e-5215ef6985ce)
6. Upgrade using the master ISO image with any custom-built image which includes this PR
7. Monitor the progress of the upgrade, the live-migratable VM should move between the two nodes and the non-live-migratable VM originally running on the control plane node should be shutoff when the control plane node is pre-draining.
8. The upgrade should end in success

*VMs with PCIe devices*

1. Prepare a 2-node Harvester cluster in v1.2.1
2. Enable `pcidevices-controller` addon
3. Enable the device you want for passthrough
4. Create a live-migratable VM with regular configs
5. Create another non-live-migratable VM with the PCI device attached
9. Start the upgrade
10. The non-live-migratable VM should be shutoff when the node it runs on is being pre-drained
11. The upgrade should end in success and the live-migratable VM is still running

*VMs with node selector*

1. Prepare a 2-node Harvester cluster in v1.2.1
2. Create a live-migratable VM with regular configs
3. Create another non-live-migratable VM with the node scheduling option set to a specific node
![SCR-20240116-phzu](https://github.com/harvester/harvester/assets/1827717/db58c1b8-b47b-4482-8ab5-ae646f930cc4)
4. Start the upgrade
5. The VM should be shutoff when the node it runs on is being pre-drained
6. The upgrade should end in success and the live-migratable VM is still running

---

For the tests against the webhook, the above three test cases could be reused. The only difference is the cluster should be installed with the master build (or custom build containing this PR), not v1.2.1. The upgrade in all three cases should be rejected by the webhook with the names of the VM which are not live-migratable.

<img width="1912" alt="image" src="https://github.com/harvester/harvester/assets/1827717/eab51473-e5c0-477d-abe7-ab2c482696ca">
